api/web3server_utils.go

```diff
@@ -0,0 +1,645 @@
+		To       string `json:"to,omitempty"`
+		Gas      string `json:"gas,omitempty"`
+		GasPrice string `json:"gasPrice,omitempty"`
+		Value    string `json:"value,omitempty"`
+		Data     string `json:"data,omitempty"`
+	}{}
+	err = json.Unmarshal(req, &callObj)
+	if err != nil {
+		return
+	}
+	if callObj.To != "" {
+		to, err = ethAddrToIoAddr(callObj.To)
+		if err != nil {
+			return
+		}
+	}
+	if callObj.From == "" {
+		callObj.From = "0x0000000000000000000000000000000000000000"
+	}
+	from, err = ethAddrToIoAddr(callObj.From)
+	if err != nil {
+		return
+	}
+	if callObj.Value != "" {
+		value, ok = big.NewInt(0).SetString(removeHexPrefix(callObj.Value), 16)
+		if !ok {
+			err = errors.Wrapf(errUnkownType, "value: %s", callObj.Value)
+			return
+		}
+	} else {
+		value = big.NewInt(0)
+	}
+	if callObj.Gas != "" {
+		gasLimit, err = hexStringToNumber(callObj.Gas)
+		if err != nil {
+			return
+		}
+	}
+	data = common.FromHex(callObj.Data)
+	return
+}
+
+func (svr *Web3Server) getLogQueryRange(fromStr, toStr string, logHeight uint64) (from uint64, to uint64, hasNewLogs bool, err error) {
+	if from, to, err = svr.parseBlockRange(fromStr, toStr); err != nil {
+		return
+	}
+	switch {
+	case logHeight < from:
+		hasNewLogs = true
+		return
+	case logHeight > to:
+		hasNewLogs = false
+		return
+	default:
+		from = logHeight
+		hasNewLogs = true
+		return
+	}
+}
+
+func loadFilterFromCache(c apiCache, filterID string) (filterObject, error) {
+	dataStr, isFound := c.Get(filterID)
+	if !isFound {
+		return filterObject{}, errInvalidFiterID
+	}
+	var filterObj filterObject
+	if err := json.Unmarshal([]byte(dataStr), &filterObj); err != nil {
+		return filterObject{}, err
+	}
+	return filterObj, nil
+}
+
+func newAPIiCache(expireTime time.Duration, remoteURL string) apiCache {
+	redisClient := redis.NewClient(&redis.Options{
+		Addr:     remoteURL,
+		Password: "", // no password set
+		DB:       0,  // use default DB
+	})
+	if redisClient.Ping(context.Background()).Err() != nil {
+		log.L().Info("local cache is used as API cache")
+		filterCache, _ := ttl.NewCache(ttl.AutoExpireOption(expireTime))
+		return &localCache{
+			ttlCache: filterCache,
+		}
+	}
+	log.L().Info("remote cache is used as API cache")
+	return &remoteCache{
+		redisCache: redisClient,
+		expireTime: expireTime,
+	}
+}
+
+type apiCache interface {
+	Set(key string, data string) error
+	Del(key string) bool
+	Get(key string) (string, bool)
+}
+
+type localCache struct {
+	ttlCache *ttl.Cache
+}
+
+func (c *localCache) Set(key string, data string) error {
+	if c.ttlCache == nil {
+		return errNullPointer
+	}
+	c.ttlCache.Set(key, data)
+	return nil
+}
+
+func (c *localCache) Del(key string) bool {
+	if c.ttlCache == nil {
+		return false
+	}
+	return c.ttlCache.Delete(key)
+}
+
+func (c *localCache) Get(key string) (string, bool) {
+	if c.ttlCache == nil {
+		return "", false
+	}
+	ret, exist := c.ttlCache.Get(key)
+	if _, ok := ret.(string); !ok || !exist {
```

Comment:
`!exist` needs to be first to check
otherwise if `exist = false`, ret = nil, it may panic